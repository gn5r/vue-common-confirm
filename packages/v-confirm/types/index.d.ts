export type Colors =
  | "primary"
  | "secondary"
  | "accent"
  | "error"
  | "info"
  | "success"
  | "warning";

export type VConfirmBtn = {
  text?: string;
  color?: Colors;
  function: FunctionConstructor;
};

import { VueConstructor } from "vue/types";
export const VConfirm: VueConstructor;
export default VConfirm;
