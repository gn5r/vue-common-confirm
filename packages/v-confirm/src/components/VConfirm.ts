// Vue
import Vue, { VNode, VNodeChildren } from "vue";

// Types
import { PropValidator } from "vue/types/options";
import { VConfirmBtn, Colors } from "@gn5r/vue-common-confirm/types";

// Component
import {
  VDialog,
  VCard,
  VCardTitle,
  VCardText,
  VCardActions,
  VSpacer,
  VToolbar,
  VToolbarTitle,
  VToolbarItems,
  VIcon,
  VBtn,
  VDivider,
} from "vuetify/lib";

export default Vue.extend({
  name: "v-confirm",
  model: {
    prop: "value",
    event: "change",
  },
  mixins: [],
  props: {
    value: Boolean,
    width: {
      type: [String, Number],
      default: 800,
    },
    closeable: Boolean,
    title: String,
    titleColor: {
      type: String,
      default: "primary",
    } as PropValidator<Colors>,
    titleIcon: String,
    message: String,
    btns: {
      type: Array,
      default: () => [],
    } as PropValidator<VConfirmBtn[]>,
  },
  data() {
    return {
      isActive: this.value,
    };
  },
  methods: {
    genCard(children: VNodeChildren): VNode {
      return this.$createElement(VCard, children);
    },
    genHeader(): VNode {
      const children = [];
      // create close button
      if (this.closeable) {
        const spacer = this.$createElement(VSpacer);
        children.push(spacer);
        const closeIcon = this.$createElement(
          VIcon,
          {
            on: {
              click: () => (this.internalValue = false),
            },
          },
          "mdi-close"
        );
        const toolbarItem = this.$createElement(VToolbarItems, [closeIcon]);
        children.push(toolbarItem);
      }
      // create title icon
      const titleIcon = this.titleIcon
        ? this.$createElement(VIcon, { props: { dark: true } }, this.titleIcon)
        : "";
      // create v-toolbar
      const toolbar = this.$createElement(
        VToolbar,
        {
          props: {
            color: this.titleColor,
            dense: true,
            dark: true,
          },
        },
        [
          this.$createElement(VToolbarTitle, [titleIcon, this.title]),
          ...children,
        ]
      );
      return this.$createElement(
        VCardTitle,
        {
          style: {
            padding: "0px",
          },
        },
        [toolbar]
      );
    },
    genContent(): VNode {
      return this.$createElement(
        VCardText,
        {
          class: ["black--text", "text-body-1"],
          style: {
            "overflow-wrap": "break-word",
            "white-space": "pre-wrap",
            padding: "16px",
          },
        },
        this.message
      );
    },
    genActions(): VNodeChildren {
      if (this.btns.length === 0) return [];
      const children = [];
      for (const btn of this.btns) {
        const button = this.$createElement(
          VBtn,
          {
            props: {
              color: btn.color,
              outlined: btn.outlined,
            },
            on: {
              click: btn.function,
            },
            class: btn.class ? btn.class : [],
          },
          btn.text
        );
        children.push(button);
      }
      return [
        this.$createElement(VDivider),
        this.$createElement(VCardActions, { class: ["justify-end"] }, [
          ...children,
        ]),
      ];
    },
  },
  computed: {
    internalValue: {
      set(val: boolean) {
        this.isActive = val;
        this.$emit("change", val);
      },
      get(): boolean {
        return this.isActive;
      },
    },
  },
  watch: {
    value(val: boolean): void {
      this.isActive = val;
    },
  },
  render(h): VNode {
    return h(
      VDialog,
      {
        props: {
          value: this.internalValue,
          scrollable: true,
          persistent: true,
          width: this.width,
        },
      },
      [this.genCard([this.genHeader(), this.genContent(), this.genActions()])]
    );
  },
});
