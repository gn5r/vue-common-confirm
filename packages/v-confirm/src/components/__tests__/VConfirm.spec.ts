import { createLocalVue, mount, MountOptions } from "@vue/test-utils"
import Vuetify from "vuetify"
import VConfirm from "@/components/VConfirm"
import { VConfirmBtn } from "@gn5r/vue-common-confirm/types"

describe("VConfirm.ts", () => {
  const localVue = createLocalVue()
  let vuetify: Vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  type Instance = InstanceType<typeof VConfirm>;
  const mountFunc = (options?: MountOptions<Instance>) =>
    mount(VConfirm, { localVue, vuetify, ...options })

  it("should render", () => {
    const wrapper = mountFunc()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render the toolbar", () => {
    const wrapper = mountFunc({
      propsData: { value: true, title: "dialog title", titleColor: "primary" }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render the close icon in toolbar", () => {
    const wrapper = mountFunc({ propsData: { value: true, closeable: true } })
    expect(wrapper.html()).toMatchSnapshot()
    const closeIcon = wrapper.find(".mdi-close")
    expect(closeIcon.exists()).toBe(true)
  })

  it("emitting change event when click the close icon", async () => {
    const wrapper = mountFunc({ propsData: { value: true, closeable: true } })
    await wrapper.find(".mdi-close").trigger("click")
    expect((wrapper.vm as any).isActive).toBe(false)
    const event = wrapper.emitted().change
    expect(event).toBeTruthy()
  })

  it("should render the message text", () => {
    const message = "test\nmessage"
    const wrapper = mountFunc({ propsData: { value: true, message: message } })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render the buttons", () => {
    const btns = [
      {
        text: "test"
      }
    ] as Array<VConfirmBtn>
    const wrapper = mountFunc({ propsData: { value: true, btns: btns } })
    expect(wrapper.html()).toMatchSnapshot()
    const cardActions = wrapper.find(".v-card__actions")
    const vbtns = cardActions.element.getElementsByClassName("v-btn")
    expect(vbtns.length).not.toBe(0)
  })

  it("should a button function is executed when click the button", async () => {
    const btns = [
      {
        text: "test",
        function: jest.fn()
      }
    ]
    const wrapper = mountFunc({ propsData: { value: true, btns: btns } })
    const btnFunction = jest.spyOn((wrapper.vm as any).btns[0], "function")
    await wrapper.find(".v-card__actions .v-btn").trigger("click")
    expect(btnFunction).toBeCalled()
  })
})
