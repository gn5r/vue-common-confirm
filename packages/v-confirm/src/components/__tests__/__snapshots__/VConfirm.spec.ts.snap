// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VConfirm.ts should render 1`] = `
<div role="dialog" class="v-dialog__container">
  <!---->
</div>
`;

exports[`VConfirm.ts should render the buttons 1`] = `
<div role="dialog" class="v-dialog__container">
  <div role="document" tabindex="0" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div class="v-dialog v-dialog--active v-dialog--persistent v-dialog--scrollable" style="transform-origin: center center; width: 800px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title" style="padding: 0px;">
            <header class="v-sheet theme--dark v-toolbar v-toolbar--dense" style="height: 48px;">
              <div class="v-toolbar__content" style="height: 48px;">
                <div class="v-toolbar__title"></div>
              </div>
            </header>
          </div>
          <div class="v-card__text black--text text-body-1" style="overflow-wrap: break-word; white-space: pre-wrap; padding: 16px;"></div>
          <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
          <div class="v-card__actions justify-end"><button type="button" class="v-btn v-btn--contained v-btn--is-elevated v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">test</span></button></div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`VConfirm.ts should render the close icon in toolbar 1`] = `
<div role="dialog" class="v-dialog__container">
  <div role="document" tabindex="0" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div class="v-dialog v-dialog--active v-dialog--persistent v-dialog--scrollable" style="transform-origin: center center; width: 800px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title" style="padding: 0px;">
            <header class="v-sheet theme--dark v-toolbar v-toolbar--dense" style="height: 48px;">
              <div class="v-toolbar__content" style="height: 48px;">
                <div class="v-toolbar__title"></div>
                <div class="spacer"></div>
                <div class="v-toolbar__items"><button type="button" class="v-icon notranslate v-icon--link mdi mdi-close theme--dark"></button></div>
              </div>
            </header>
          </div>
          <div class="v-card__text black--text text-body-1" style="overflow-wrap: break-word; white-space: pre-wrap; padding: 16px;"></div>
          <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
          <div class="v-card__actions justify-end"></div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`VConfirm.ts should render the message text 1`] = `
<div role="dialog" class="v-dialog__container">
  <div role="document" tabindex="0" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div class="v-dialog v-dialog--active v-dialog--persistent v-dialog--scrollable" style="transform-origin: center center; width: 800px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title" style="padding: 0px;">
            <header class="v-sheet theme--dark v-toolbar v-toolbar--dense" style="height: 48px;">
              <div class="v-toolbar__content" style="height: 48px;">
                <div class="v-toolbar__title"></div>
              </div>
            </header>
          </div>
          <div class="v-card__text black--text text-body-1" style="overflow-wrap: break-word; white-space: pre-wrap; padding: 16px;">test
            message</div>
          <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
          <div class="v-card__actions justify-end"></div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`VConfirm.ts should render the toolbar 1`] = `
<div role="dialog" class="v-dialog__container">
  <div role="document" tabindex="0" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div class="v-dialog v-dialog--active v-dialog--persistent v-dialog--scrollable" style="transform-origin: center center; width: 800px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title" style="padding: 0px;">
            <header class="v-sheet theme--dark v-toolbar v-toolbar--dense primary" style="height: 48px;">
              <div class="v-toolbar__content" style="height: 48px;">
                <div class="v-toolbar__title">dialog title</div>
              </div>
            </header>
          </div>
          <div class="v-card__text black--text text-body-1" style="overflow-wrap: break-word; white-space: pre-wrap; padding: 16px;"></div>
          <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
          <div class="v-card__actions justify-end"></div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
